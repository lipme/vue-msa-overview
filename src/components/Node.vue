<template>
<g :class="classAttribute" :transform="transformAttribute" style="transition:transform 0.5s ease-out">
    <circle :r="type === 'inner' ? size/3 : size/2"
     :stroke-width="strokeWidth"
     >
      <!-- <animate attributeType="XML" attributeName="r" from="0" :to="type === 'inner' ? size/3 : size/2" dur="0.5s" /> -->
     </circle>
    <text v-if="showLabel" :dx="optionsLabel.x"
    :dy="optionsLabel.y" :text-anchor="optionsLabel['text-anchor']"
    :font-size="optionsLabel['font-size']"
    :transform="optionsLabel.transform" >
    {{this.label}}
    <animate attributeType="XML" attributeName="font-size" from="0" :to="optionsLabel['font-size']" dur="0.8s" />
    </text></g>
</template>

<script>
export default {
  props: {
    size: {
      type: Number,
      default: 6
    },
    x: {
      type: Number,
      default: 0
    },
    y: {
      type: Number,
      default: 0
    },
    label: {
      type: String,
      default: null
    },
    type: {
      type: String,
      default: ''
    },
    circular: {
      type: Boolean,
      default: false
    },
    id: {
      type: String,
      default: null
    },
    selected: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    strokeWidth () {
      return this.size / 5
    },
    textClass () {
      if (this.circular && this.x > 180) {
        return 'reverse'
      }
      return 'normal'
    },
    showCircle () {
      return this.type !== 'inner'
    },
    classAttribute () {
      console.log('change attribute')
      return 'node ' + this.type + ' ' + this.textClass + (this.selected ? ' selected' : '')
    },
    transformAttribute () {
      if (this.circular === false) {
        return 'translate(' + this.y + ' ' + this.x + ')'
      } else {
        return 'rotate(' + (this.x - 90) + ')translate(' + this.y + ')'
      }
    },
    optionsLabel () {
      // const transform = ''
      let anchor = 'start'
      let x = 1.5 * this.size

      if (this.circular && this.x > 180) {
        // transform = 'rotate(180)'
        anchor = 'end'
        x = -1.5 * this.size
      }

      return this.type === 'inner'
        ? { x: x, y: -this.size / 3, 'text-anchor': anchor, 'font-size': this.size + 3 + 'px' }
        : { x: x, y: this.size / 3, 'text-anchor': anchor, 'font-size': this.size + 3 + 'px' }
    },
    showLabel () {
      return this.label !== null && this.label !== '' && this.type !== 'root'
    }

  }
}
</script>

<style scoped>

circle {
  transition: all 0.5s;
  opacity: 100%;
  stroke:darkblue;
  fill:steelblue
}

.root circle {
  stroke: yellowgreen;
  fill: greenyellow;
}

.inner circle {
  opacity: 0%;
  fill:lightsalmon;
  stroke: coral;
}

.selected circle {
  opacity: 100%;
  transition: all 0.5s;
  fill: red;
  stroke: brown;
  transform: scale(1.5);
}

.selected text {
  fill: red;
  font-weight: bold;
  transform: scale(1.5);
}

circle:hover {
  transition: all 0.5s;
  transform: scale(1.5);
  opacity: 100%;
}

text:hover {
  transition: all 0.5s;
  transform: scale(1.5);
}

.reverse text:hover {
  transition: all 0.5s;
  transform: scale(1.5) rotate(180deg);
}

.reverse text {
  transform: rotate(180deg);
}

text {
  fill: black;
  font-family: 'Helvetica Neue, Helvetica, sans-serif'
}

</style>
