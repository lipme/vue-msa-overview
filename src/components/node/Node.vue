<template>
<g :transform="transformAttribute">
    <circle v-if="!collapsed" :class="{transparent: !show}" :r="size/2"
     :stroke-width="strokeWidth" :stroke="strokeColor" :fill="fill"
     >
      <animateTransform attributeName="transform"
             type="scale"
             from="0 0"
             to="1 1"
             begin="0s"
             dur="1s"
        />
     </circle>
       <polygon v-else :points="triangleCoords" :fill="fill" :stroke-width="strokeWidth" :stroke="strokeColor"/>
    </g>
</template>

<script>
export default {
  name: 'node',
  props: {
    size: {
      type: Number,
      default: 6
    },
    x: {
      type: Number,
      default: 0
    },
    y: {
      type: Number,
      default: 0
    },
    circular: {
      type: Boolean,
      default: false
    },
    show: {
      type: Boolean,
      default: true
    },
    strokeColor: {
      type: String,
      default: 'darkblue'
    },
    fill: {
      type: String,
      default: 'steelblue'
    },
    collapsed: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    strokeWidth () {
      return this.size / 5
    },
    transformAttribute () {
      if (this.circular === false) {
        return 'translate(' + this.y + ' ' + this.x + ')'
      } else {
        return 'rotate(' + (this.x - 90) + ')translate(' + this.y + ')'
      }
    },
    triangleCoords () {
      const coords = [
        { x: 0, y: 0 },
        {
          x: this.size * 2,
          y: -this.size * 2
        },
        {
          x: this.size * 2,
          y: this.size * 2
        }
      ]
      return `${coords[0].x} ${coords[0].y},${coords[1].x} ${coords[1].y}, ${coords[2].x} ${coords[2].y}`
    }
  }
}
</script>

<style scoped>

.transparent {
  opacity:0%;
}

circle:hover,polygon:hover {
  transition: all 0.5s;
  transform: scale(2);
  opacity: 100%;
}

</style>
