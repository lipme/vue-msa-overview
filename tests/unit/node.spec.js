import { shallowMount } from '@vue/test-utils'
import Node from '@/components/Node.vue'

describe('Node component', () => {
  it('root node renders correctly', () => {
    const comp = shallowMount(Node, {
      propsData: { type: 'root', size: 5 }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('inner node renders correctly', () => {
    const comp = shallowMount(Node, {
      propsData: { type: 'inner', size: 3 }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('leave node renders correctly', () => {
    const comp = shallowMount(Node, {
      propsData: { type: 'inner', size: 3 }
    })

    expect(comp.element).toMatchSnapshot()
  })
})
