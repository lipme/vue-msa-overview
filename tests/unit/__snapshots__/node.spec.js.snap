// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node component collapsed mode set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <polygon
    fill="steelblue"
    points="0 0,9 -9, 9 9"
    stroke="darkblue"
    stroke-width="1.2"
  />
</g>
`;

exports[`Node component default node renders correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`Node component fill color set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="black"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`Node component node position in circular mode set correctly 1`] = `
<g
  transform="rotate(-90)translate(0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`Node component node size set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="5"
    stroke="darkblue"
    stroke-width="2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`Node component node x set correctly 1`] = `
<g
  transform="translate(0 100)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`Node component node y set correctly 1`] = `
<g
  transform="translate(100 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`Node component show mode set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class="transparent"
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`Node component stroke color set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="black"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;
